### DISCUSSION TITLE
ISR in multi cluster nodes environment with CDN in front

### DISCUSSION THREAD

**Discussion by alfrescian** [vercel/next.js#29291] [Help]
I spent time reading some blog posts, issue tickets & discussions threads here but I'm still a bot lost regarding ISR  feature when running multiple NextJS cluster nodes/instances in a scale-out scenario.
(we need multiple NextJS nodes to also serve SSR pages that create a higher server load)

let's assume the following scenario:
- we have a ISR page /product/[key] with revalidate=60 & fallback=blocking
- we have 2 NextJS cluster nodes (node_1, node_2)
- A CDN in front fo of our cluster

Assume the product data of product.key=A is changed after initial SSG build 
1. user1 hits /product/A after 2 minutes
2. user1 gets stale page from CDN (version 1 from node 1 =  version1:1)
3. CDN revalidates the page async in the background using node_1
4. node_1 sends stale page (version 1:1) to CDN & rebuild the page with current data -> version 2:1.
5. after2 minutes user2 hits  /product/A
6. user2 gets stale page from CDN (version 1:1)
7. CDN revalidates the page async in the background using node_1
8. node_1 sends stale page (version 2:1) to CDN & does not rebuild as staticProps are unchanged
9. after2 minutes user3 hits  /product/A
10. user3 gets stale page from CDN (version 2 - finally a user get's the current data)
11. CDN revalidates the page async in the background using node_2
12. node_2 sends stale page (version 2:1 ... yes the old one!) to CDN & rebuild the page with current data -> version 2:2. 
13. then user3 hits /product/A again
14. user3 gets page from CDN (version 2:1 -  so he gets the old version again)

Is that flow correct? 
If so, I assume that  - without any additional work* - ISR is pretty unusable if yo run nextJS is a scale-out scenario.
the CDN revalidate-dance makes it even worse because the end user may get a pretty old version of a page (`currentVersion - numberOfNodes +1`)

Am I correct here did I oversee/mixup sth?

*lets discuss "without any additional work":
thing we could do: 
A. share the _next/static folder between cluster nodes, so that no node has a page with version < currentVersion
B. purge cache on all cluster nodes as soon as product-data is changed, so that the revalidate requests from CDN is a `fallback:blocking` request and the updated page is generated in blocking mode.

**Comment by michaelschufi** [comment]
Hi @alfrescian 

Have you found a solution to this in the meantime? I'm facing the same challenges at the moment. Thinking of making my own ISR cache using a shared Redis for example, but if there's a solution compatible with Next.js' ISR, that would be great.

**Reply by sigdestad** [reply]
I guess the architectural problem is using a CDN to cache html. It's generally an anti-pattern when you have an application generating the html "dynamically" behind the CDN. 

As Next.js uses static generation it is as fast as the CDN. If you need global performance, consider deploying next servers globally I guess?

**Reply by alfrescian** [reply]
hi - we solved the problem by :
- using reasonable ISR Revalidate timings (no every page needs to revalidated every 60s...).
- using ISR revalidate API to revalidate pages if sth in our data-sources (eg. CMS) changes
- fully switch to ISR & CSR and no more usage of SSR

all of this in combination enables us to run with a single node.

> I guess the architectural problem is using a CDN to cache html.

Yub, using a CDN makes the origin problem a bit harder, but you'll face the same problem w/o a CDN as well. As soon as you run a multi-node cluster on your own with ISR you will have to think about "is delivering outdated html a problem for my use-case".
If your answer is a "yes", then you should consider to share your ISR cache updates (redis etc.) or deploy on a platform that solves the issue (eg. vercel)
